libmamba/include/mamba/core/solver.hpp

```diff
@@ -67,17 +68,21 @@ namespace mamba
         void add_pins(const std::vector<std::string>& pins);
         void set_flags(const std::vector<std::pair<int, int>>& flags);
         void set_postsolve_flags(const std::vector<std::pair<int, int>>& flags);
-        bool is_solved() const;
-        bool solve();
-        std::string problems_to_str() const;
-        std::vector<std::string> all_problems() const;
-        std::vector<MSolverProblem> all_problems_structured() const;
-        std::string all_problems_to_str() const;
-
-        const std::vector<MatchSpec>& install_specs() const;
-        const std::vector<MatchSpec>& remove_specs() const;
-        const std::vector<MatchSpec>& neuter_specs() const;
-        const std::vector<MatchSpec>& pinned_specs() const;
+        [[nodiscard]] bool try_solve();
```

Comment:
how about we change the interface to return a `tl::expected` instead? Then we can keep a single function that we can test for solver problems?

Reply:
I don't know. When I tried to use them I found they don't compose nicely. Here there would also not be any outcome, so does it make sense to have a `tl::expected<void, Error>`?

I also introduce `must_solve` that throws an exception because some code used `solve` without checking the outcome.
Since I would not know what to do in those cases, the exception seems like the natural first step.